File: 406_QueueReconstructionByHeight.cpp
//24.34%
class Solution {
public:
    vector<pair<int, int>> reconstructQueue(vector<pair<int, int>>& people) {
        if(people.empty())
			return {};
		
		unordered_map<int, vector<pair<int, int> > > peopleTable;
		vector<int> heights;
		vector<pair<int, int> > res;
		for(int i = 0; i < people.size(); ++i){
			if(peopleTable.find(people[i].first) == peopleTable.end()){
				peopleTable.insert(make_pair(people[i].first, vector<pair<int, int> >(1, make_pair(people[i].second, i))));
				heights.push_back(people[i].first);
			}else
				peopleTable[people[i].first].emplace_back(people[i].second, i);
		}
		
		sort(heights.begin(), heights.end(), greater<int>());
		for(auto height : heights){
			sort(peopleTable[height].begin(), peopleTable[height].end());
			for(auto& p : peopleTable[height]){
				res.insert(res.begin() + p.first, people[p.second]);
			}
		}
		return res;	
    }
};


vector<pair<int, int>> reconstructQueue(vector<pair<int, int>>& people) {
    auto comp = [](const pair<int, int>& p1, const pair<int, int>& p2)
                    { return p1.first > p2.first || (p1.first == p2.first && p1.second < p2.second); };
    sort(people.begin(), people.end(), comp);
    vector<pair<int, int>> res;
    for (auto& p : people) 
        res.insert(res.begin() + p.second, p);
    return res;
}

File: 406_QueueReconstructionByHeight.py
##54.57%

#Pick out tallest group of people and sort them in a subarray (S). Since there's no other groups of people taller than them, therefore each guy's index will be just as same as his k value.
#For 2nd tallest group (and the rest), insert each one of them into (S) by k value. So on and so forth.
#E.g.
#input: [[7,0], [4,4], [7,1], [5,0], [6,1], [5,2]]
#subarray after step 1: [[7,0], [7,1]]
#subarray after step 2: [[7,0], [6,1], [7,1]]
#...

#It's not the most concise code, but I think it well explained the concept.

class Solution(object):
    def reconstructQueue(self, people):
        if not people: return []

        # obtain everyone's info
        # key=height, value=k-value, index in original array
        peopledct, height, res = {}, [], []
        
        for i in xrange(len(people)):
            p = people[i]
            if p[0] in peopledct:
                peopledct[p[0]] += (p[1], i),
            else:
                peopledct[p[0]] = [(p[1], i)]
                height += p[0],

        height.sort()      # here are different heights we have

        # sort from the tallest group
        for h in height[::-1]:
            peopledct[h].sort()
            for p in peopledct[h]:
                res.insert(p[0], people[p[1]])

        return res
		
		
def reconstructQueue(self, people):
    people.sort(key=lambda (h, k): (-h, k))
    queue = []
    for p in people:
        queue.insert(p[1], p)
    return queue